pool.go

```diff
@@ -239,46 +253,60 @@ func (p *Pool) Start(closeFunc func()) {
 			}
 			log.Info("finish closing pool")
 
+			// wait for all on-fly retryable jobs are inserted to db
+			p.retryableWaitGroup.Wait()
+
 			// send signal to stop the program
-			p.stop <- struct{}{}
+			close(p.stop)
 			return
 		}
 	}
 }
 
 func (p *Pool) Stats() Stats {
 	return Stats{
-		PendingQueue: len(p.PrepareJobChan),
-		Queue:        len(p.JobChan),
+		RetryableQueue: len(p.RetryJobChan),
+		Queue:          len(p.JobChan),
+		RetryingJob:    atomic.LoadInt32(&p.numberOfRetryingJob),
 	}
 }
 
 func (p *Pool) PrepareRetryableJob(job JobHandler) {
+	// if pool is closed, try update job to db
+	if p.isClosed.Load().(bool) {
+		log.Info("pool closed, update retrying job to database")
+		p.updateRetryingJob(job)
+	}
 	dur := time.Until(time.Unix(job.GetNextTry(), 0))
 	if dur <= 0 {
 		return
 	}
+	atomic.AddInt32(&p.numberOfRetryingJob, 1)
+	p.retryableWaitGroup.Add(1)
+
+	defer func() {
+		p.retryableWaitGroup.Done()
+		atomic.AddInt32(&p.numberOfRetryingJob, -1)
+	}()
+
 	timer := time.NewTimer(dur)
 	select {
 	case <-timer.C:
-		p.PrepareJobChan <- job
+		p.Enqueue(job)
 	case <-p.ctx.Done():
+		log.Info("pool closed, update retrying job to database")
 		p.updateRetryingJob(job)
 	}
 }
 
-// PrepareJob saves new job to database
-func (p *Pool) PrepareJob(job JobHandler) error {
-	if job == nil {
-		return errors.New("job is nil")
-	}
-	return nil
-}
-
 func (p *Pool) updateRetryingJob(job JobHandler) {
 	if job == nil {
 		return
 	}
+
+	p.lock.Lock()
```

Comment:
I think this lock is redundant. AFAIK, 2 writes to same row in postgresql are serialized, aren't they?

Reply:
we use this to lock the pool, not the db, so that the pool cannot exit without finishing this function

Reply:
For retry job, I think the waitgroup is enough, I'm not sure about the failed job flow. However, I cannot see how this lock can prevent pool from exiting.

Reply:
There can be some goroutines still calling to `Enqueue` or `RetryJob` and panic because of "send to close channel" on both `Jobchan` and `RetryJobChan` will happen. This time the pool may already closed and waitgroup for on-the-fly retry-able jobs may be all done. The pool then exits without waiting these events. Calling lock will make sure it cannot close until these events are stored to db